### DISCUSSION TITLE
Dreamcast keyboard

### DISCUSSION THREAD

**Discussion by ruenes1** [darthcloud/BlueRetro#332] [Q&A] [answered]
I've built a two player adaptor for the DC and so far it has been working great with playstation gamepads. 

Now I've tried to also connect a bluetooth keyboard to P1. For this, I've set the 'output1' to 'Keyboard' (and 'output2' to gamepad (no rumble)), and also loaded the preset  'default gamepad/keyboard' to output1.

The keyboard seems to connect fine with the adapter (the external led in the adapter goes from pulsing to off when the keyboard is paired), but I get no keyboard response in the dreamcast (for keyboard-enabled games such as the typing of the dead). 

This is the serial log I got when connecting to the keyboard. I'm using FW version 1.6.3.

[blueretro.txt](https://github.com/darthcloud/BlueRetro/files/9071741/blueretro.txt)

Any tips? Thanks for your time!

**Comment by darthcloud** [comment]
I think the problem is that this keyboard got 2 reports that get identified as keyboards reports. The second one probably for special features, but because its the last one it endup taking precedence over the first and real ones.

I will need to fix that! Thanks for providing log!

**Reply by ruenes1** [reply]
Great, thanks! Yes, the keyboard is a 60% with a second layer of key functions for extra characters and modes, so maybe that was the problem. It also supports multiple pre-stored bluetooth bindings.

I will be happy to test any fix!

**Comment by darthcloud** [comment, marked as answer]
Try this firmware:
[v1.6.3_hid_report_fix.zip](https://github.com/darthcloud/BlueRetro/files/9086458/v1.6.3_hid_report_fix.zip)

**Reply by ruenes1** [reply]
I've just tried it and it works great! I'm able to pair both the keyboard (P1) and a DS4 (P2) and both work. 

The only issue I found (but probably has nothing to do with the adapter) is that, in some games, if I type fast or press the keys for a too small amount of time, some of them do not register. This happens, for example, in the Typing of the dead, whose goal is to type fast. I tried to have the keyboard just side by side with the receiver but didn't changed this. 

The keyboard, when paired with other devices (windows pc, android), works fine and does not lose inputs no mater how fast I type. This neither happended in other DC games, so maybe is a per-game issue. Unfortunately, I don't have other bluetooth keyboards that I could test.

**Reply by darthcloud** [reply]
I'll give Typing of the dead a try when I find time with one of mine and report back.

**Comment by rpf16rj** [comment]
Hi, how are you? I hope fine!

I am try use keyboard on a Dreamcast. I made a adapter and he work fine with a bluetooth PS4 joystick.
When i try use in keyboard mode (i setted the mode to Keyboard and preset Gamepad;Keyboard), the kayboard show connected on Dreamcast but no response of keys. 

I tested that this version of fw in this thread and the lasted 1.8.3.

In another devices, the both keyboards works fine.

I tested with two models of keyboards bt, a Microsoft Mobile Keyboard 6000 and a generic bt keyboard chinese. I atached the debug log below. 

Thanks for you time. 

[bt_keyboard_microsoft_mobile_6000.txt](https://github.com/darthcloud/BlueRetro/files/11479638/bt_keyboard_microsoft_mobile_6000.txt)
[bt_keyboard_chinese.txt](https://github.com/darthcloud/BlueRetro/files/11479639/bt_keyboard_chinese.txt)
